### ISSUE TITLE
DOC: lazy transpose function behavior is undocumented

### ISSUE THREAD

**Issue by olafx** [numpy/numpy#20508]
### Issue with current documentation:

I discovered that the transpose function (`transpose(array)`) just reverses the dimensions and changes from row major ('contiguousarray') storage to column major ('fortranarray') storage. Nothing wrong with this, is smart, sort of a lazy evaluation.
It only actually moves data when I use `ascontiguousarray(transpose(array))`, which again makes sense, no problem.
The problem is, this is undocumented. It says nothing [here](https://numpy.org/doc/stable/reference/generated/numpy.transpose.html) about changing from the standard contiguous array layout to Fortran array layout. This isn't obvious at all that it would do this, and can make for unexpected results, because NumPy row major and column major arrays aren't interchangeable just because functions work with both. (I discovered it due to writing raw data and found unexpected behavior, namely that transpose did nothing, so proof is in the pudding.)

### Idea or request for content:

No mention of changing memory layout from contiguous to fortran in `numpy.reverse` as a means of lazy evaluation.

**Comment by bashtage** [comment]
It is sort of hiding in this statement: "A view is returned whenever possible". A view implies that data is not copied which then requires what you have noticed.

**Comment by melissawm** [comment]
@olafx do you have a suggestion for re-wording in the docs that would make this clearer? Happy to push this forward if that's the case.

**Comment by seberg** [comment]
NumPy's guarantees on memory order are mostly best-effort.  Almost nothing documents that a function tries to conserve memory order on its output, or fails to do so, but might in the future.

I doubt it is helpful to note that on a per-function level, and I wouldn't be surprised if it is already mentioned in the memory-layout section of the documentation.  I am sure there could be some clarifications somewhere, or just links to the details, though.

**Comment by olafx** [comment]
If memory order isn't documented in general, maybe best to close this issue and not mention it for transpose either. But otherwise I'd reword like this: "For an array a with two axes, transpose(a) gives the matrix transpose. If a has C order, transpose(a) will have Fortran order, and vice versa." This follows the wording in the `ascontiguousarray` and `asfortranarray` documentation. Probably best to add `ascontiguousarray` and `asfortranarray` in 'see also' to let readers know what that means.
